More License Detection changes

AyanSinhaMahapatra opened the pull request:
* Creates a top-level `licenses` attribute with unique license detections
* makes the licenses-reference default, and instead adds a --no-licenses-reference option 


### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Add codebase level unique license detections

- Add a new codebase level attribute `licenses`
- Add a new resource level attribute `for_licenses`
- Add unique license detections from files and packages
  in the top level attribute `licenses` and this is the
  usual `license_expression`, `detection_log` and `matches`
  and additionally an `occurance_count` and a `identifier`
  which is an UUID generated from the content of the matches
  in the detection.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expectations with top level licenses

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Make license references default

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regen test expectations after default license references

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on tests/cluecode/data/plugin_email_url/urls-threshold.expected.json:
```diff
@@ -1,4 +1,6 @@
 {
+  "license_references": [],
+  "rule_references": [],
```
Rule references is not an obvious, non-ambiguous name for a top level attributes. It would have to be `license_rule_references` or in any case something that very explicitly is about licenses

pombredanne replied:
Also for non-license CLI tests use the --no-license-references option this way the expectation will not change

AyanSinhaMahapatra replied:
Yes changing it to `license_rule_references`, that's much better.

pombredanne commented on tests/licensedcode/data/plugin_license/text/scan.expected.json:
```diff
@@ -21,69 +183,16 @@
       ],
       "license_clues": [],
       "percentage_of_license_text": 100.0,
+      "for_licenses": [
```
what is a licenses in this context? is this a license detection? if so we have to be explicit ... and also I wonder if we could come with an id that's a little more friendly than a UUID? may be an expression cleaned with the `commoncode.text.python_safe_name` function and padding a dash and the UUID? And if this is about license detections, then IMHO the top level attribute should be also `license_detections` and not a plain licenses

AyanSinhaMahapatra replied:
Implementing this.

pombredanne commented on tests/licensedcode/data/plugin_license/text/scan.expected.json:
```diff
@@ -107,39 +216,16 @@
-                  "spdx_url": "https://spdx.org/licenses/FSFAP"
-                }
-              ]
+              "rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/fsf-ap.LICENSE"
```
To have reasonably stable URLs will need to update the models to ensure that we never reuse existing rule ids and never rename rules either.... we need to move to a soft-delete mode like with licenses... or could we publish the rules on the license DB too?

AyanSinhaMahapatra replied:
Yes! we should ensure we don't change rule names (will open an issue for that?), we also have this issue for rules in licenseDB: https://github.com/nexB/scancode-licensedb/issues/3

pombredanne commented on tests/licensedcode/data/plugin_license/text/scan.expected.json:
```diff
@@ -1,4 +1,166 @@
+      "is_license_intro": false,
+      "rule_length": 13,
+      "rule_relevance": 100,
+      "matched_text": "SPDX-License-Identifier: ((GPL-2.0 WITH Linux-syscall-note) OR Linux-Openib) */"
```
Is this really it? is this the full rule text?... IMHO the matched text is NOT something related top a rule reference, this is an attribute of a license match, e.g., at the intersection between a rule and a scanned file.

AyanSinhaMahapatra replied:
yes, this belongs in the match details. Moving it there.

pombredanne commented on tests/licensedcode/data/plugin_license/text/scan.expected.json:
```diff
@@ -1,4 +1,166 @@
+  ],
+  "rule_references": [
+    {
+      "license_expression": "gpl-2.0 WITH linux-syscall-exception-gpl OR linux-openib",
```
IMHO here we should just dump the rule, as is. The order should be that of https://github.com/nexB/scancode-toolkit/blob/a410f6ebb93a9b22343d7737897b61f3cd2a6ac3/src/licensedcode/models.py#L1773 .... but that order is not correct and should be updated to align exactly with the order of a dump, and the text should come last.

AyanSinhaMahapatra replied:
addressed. Closing this.

pombredanne reviewed (changes requested):
Thanks!
See some comments inline for your review.

AyanSinhaMahapatra committed:
Make licenses referance default in license detection

* removes the `--licenses-reference` CLI option and plugin
* the license_references and license_rule_references attributes
  are now default with `--license` option and the same has been removed
  from match level data to avoid data duplication.
* tests are reorganized and files renamed

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Refactor license detection to use rehydrated objects

* Due to license references being default, reference data isn't
  inlined anymore, so we need to use the cache to get this data
  and also rehydrate them into objects to be able to post process
  license related info.
* Use license objects wherever possible instead of mappings.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regenerate test expectations

* rename top level attirbute `licenses` -> `license_detections`
* rename top level attribute `rule_references` to `license_rule_references`
* include license_expression in identifier
* use the correct spelling for occurrance
* include matched_text in matches instead of reference data
* file level attribute `for_licenses` changed to `for_license_detections`
* fix bugs including license rule references correctly
* make license references default to `--licenses`

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Don't remove license references in package license detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Reorder codebase and resource attributes

* Reorder codebase and resource level attributes
* replace `#` in license detection identifier with `-`
* regenerate test expectations
* reorder license rule references attributes
* add rule text to license rule references data

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update license rules reference data

* Add rule text to reference data
* Add rule url to reference data
* make rule references unique
* reorder rule references data
* regenerate test expectations

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
#3150 now has all the changes from this PR.

pombredanne commented on tests/licensedcode/data/plugin_license/sqlite/sqlite.expected.json:
```diff
@@ -1,4 +1,60 @@
+      "owner": "SQLite",
+      "homepage_url": "https://sqlite.org/src/artifact/df5091916dbb40e6",
+      "is_builtin": true,
+      "spdx_license_key": "blessing",
```
I wonder if we should not force including all the attributes even if empty? That's been the standard for how we report things in SCTK so far

pombredanne replied:
This can wait for later

AyanSinhaMahapatra replied:
@pombredanne do you mean the `is_builtin` flag? That is the only attribute we are adding whenever there are extra licenses installed/added to the index, our previous discussion was we add other attributes too based on CLI options/other optional elements. But apart from this all other attributes are always added to the reference. Note that we have more attributes added on tge release prep branch.

pombredanne replied:
I meant all the attributes that are empty. For instance in https://github.com/nexB/scancode-toolkit/blob/f1cc917e002d4269869ffefa47acb1bdfb3d2d56/src/licensedcode/models.py#L393 we do not report these

pombredanne replied:
I entered https://github.com/nexB/scancode-toolkit/issues/3182

pombredanne reviewed (approved):
LGTM! merging!
